### PR TITLE
gh-108901: Add `bound_arg` to `Signature.from_callable()` and `signature()`

### PR DESCRIPTION
**Pull request by sobolevn**
* Issue: gh-108901

----
📚 Documentation preview 📚: https://cpython-previews--116559.org.readthedocs.build/

### CODE REVIEW

File: Doc/library/inspect.rst [hunk 1]
```diff
@@ -693,6 +693,10 @@ function.
    .. versionchanged:: 3.10
       The *globals*, *locals*, and *eval_str* parameters were added.
 
+   .. versionchanged:: 3.13
+      The *skip_bound_arg* parameter was added.
+      Pass ``False`` to keep the ``self`` parameter in a signature.
```
Comment (vstinner) [inline Doc/library/inspect.rst]: Please move this sentence ("Pass...") from this note to the doc body. Or rephrase it to something like:

If *skip_bound_arg* is false, keep `self` parameter in the signature.

File: Doc/whatsnew/3.13.rst [hunk 2]
```diff
@@ -361,6 +361,14 @@ and only logged in :ref:`Python Development Mode <devmode>` or on :ref:`Python
 built on debug mode <debug-build>`.
 (Contributed by Victor Stinner in :gh:`62948`.)
 
+inspect
+-------
+
+* Add *skip_bound_arg* parameter to :func:`inspect.Signature.from_callable`
+  and :func:`inspect.signature`, pass ``False``
+  to keep the ``self`` parameter in a signature.
```
Comment (vstinner) [inline Doc/whatsnew/3.13.rst]:
```suggestion
  and :func:`inspect.signature`: keep the ``self`` parameter
  in the signature if *skip_bound_arg* is False.
```

File: Lib/test/test_inspect/test_inspect.py [hunk 3]
```diff
@@ -4622,6 +4622,65 @@ class D2(D1):
...
+                        self.assertEqual(str(sig), text_sig)
+
+    def test_signature_skip_bound_arg_function(self):
+        def compare(self: object, other: object) -> bool: ...
+
+        for func in (inspect.signature, inspect.Signature.from_callable):
+            with self.subTest(func=func):
+                sig = func(compare, skip_bound_arg=False)
+                self.assertEqual(str(sig),
+                                 '(self: object, other: object) -> bool')
```
Comment (vstinner) [inline Lib/test/test_inspect/test_inspect.py]: Can you please test also with skip_bound_arg=True?

File: Lib/test/test_inspect/test_inspect.py [hunk 4]
```diff
@@ -4622,6 +4622,67 @@ class D2(D1):
...
+    def test_signature_skip_bound_arg_function(self):
+        def compare(self: object, other: object) -> bool: ...
+
+        for func in (inspect.signature, inspect.Signature.from_callable):
+            with self.subTest(func=func):
+                sig1 = func(compare, skip_bound_arg=False)
+                sig2 = func(compare, skip_bound_arg=True)
+                self.assertEqual(str(sig1),
+                                 '(self: object, other: object) -> bool')
+                self.assertEqual(sig1, sig2)
```
Comment (vstinner) [inline Lib/test/test_inspect/test_inspect.py]: I don't understand: skip_bound_arg has no effect? What's the purpose of the test in this case?

Comment (sobolevn) [reply Lib/test/test_inspect/test_inspect.py]: This test ensure that `self` is not removed from a function, only from methods. I think that this test is important to ensure that `skip_bound_arg` does not change that.

Comment (vstinner) [reply Lib/test/test_inspect/test_inspect.py]: Ok, but I would like to see the real effect on a method where the flag changes the signature. Is there a test like that?

Maybe in the same test, add a class with a method, and have a similar test on the method?

Comment (sobolevn) [reply Lib/test/test_inspect/test_inspect.py]: I've added the change demo for methods 👍

File: Lib/test/test_inspect/test_inspect.py [hunk 5]
```diff
@@ -4622,6 +4622,122 @@ class D2(D1):
...
+                My.decorated: (
+                    '(self, arg: int) -> None'
+                    if follow_wrapped else
+                    '(*args, **kwargs) -> None'
+                ),
+                My.st: '(arg1: str) -> bool',
+                My().st: '(arg1: str) -> bool',
+            }
+
+            for func in (
```
Comment (vstinner) [inline Lib/test/test_inspect/test_inspect.py]: I suggest to rename `func` to `get_signature` or `signature_func`.

File: Lib/test/test_inspect/test_inspect.py [hunk 6]
```diff
@@ -4622,6 +4622,122 @@ class D2(D1):
...
+                inspect.signature,
+                inspect.Signature.from_callable,
+            ):
+                for fixture, text_sig in unbound_sigs.items():
```
Comment (vstinner) [inline Lib/test/test_inspect/test_inspect.py]: I'm not sure what `fixture` stands for here. signature() first parameter is called `callable`, I suggest to reuse this name here.

File: Lib/test/test_inspect/test_inspect.py [hunk 7]
```diff
@@ -4622,6 +4622,122 @@ class D2(D1):
...
+        class My:
+            def method(self, arg: int) -> None: ...
+            @decorator
+            def decorated(self, arg: int) -> None: ...
+            @classmethod
+            def cl(cls, arg2: str) -> None: ...
+            @staticmethod
+            def st(arg1: str) -> bool: ...
+
+        for follow_wrapped in (True, False):
```
Comment (vstinner) [inline Lib/test/test_inspect/test_inspect.py]: This function has too many nested loops for my little brain :-( Can you rename `test_signature_skip_bound_arg_method` to `check_signature_skip_bound_arg_method` and have two methods:

```
def test_method_skip_bound_arg_follow_wrapped(self):
    self.check_signature_skip_bound_arg_method(follow_wrapped=True)

def test_method_skip_bound_arg_no_follow_wrapped(self):
    self.check_signature_skip_bound_arg_method(follow_wrapped=True)
```

Well. Something like that.

**Comment by vstinner** [comment]
By the way, `skip_bound_arg` parameter name is a little bit long. Maybe you can rename it to `bound_arg`.

@pitrou: Do you have an opinion on the parameter name? :-)

**Comment by sobolevn** [comment]
I would prefer to have `skip_bound_arg` as a name, because it has a history: https://github.com/python/cpython/blob/4e5df2013fc29ed8bdb71572f1d12ff36e7028d5/Lib/inspect.py#L1404-L1413

**Comment by vstinner** [comment]
**skip**_bound_arg=**False** is a double negation, and I'm always confused by double negation :-( bound_arg=**True** is more straightfoward to me.

Previously, the argument was private, no?

**Comment by sobolevn** [comment]
Yes, it was private. It only existed on protected `_signature_from_callable` helper.

File: Doc/library/inspect.rst [hunk 8]
```diff
@@ -665,6 +665,11 @@ function.
    Accepts a wide range of Python callables, from plain functions and classes to
    :func:`functools.partial` objects.
 
+   If *follow_wrapped* is ``False`` *callable* will not be unwrapped
+   (``callable.__wrapped__`` will not be used to unwrap decorated callables).
+
+   If *skip_bound_arg* is ``False``, keep ``self`` parameter in a signature.
```
Comment (skirpichev) [inline Doc/library/inspect.rst]: Maybe it should be obvious to readers, e.g. from the parameter name, but you could expand this a bit.  The question is: signature of what?  E.g.:
```pycon
>>> class Spam:
...     def foo(self, bar): ...
... 
>>> import inspect
>>> inspect.signature(Spam.foo)
<Signature (self, bar)>
>>> inspect.signature(Spam.foo, skip_bound_arg=False)
<Signature (self, bar)>
>>> inspect.signature(Spam().foo)
<Signature (bar)>
>>> inspect.signature(Spam().foo, skip_bound_arg=False)
<Signature (self, bar)>
```
```suggestion
   If *skip_bound_arg* is ``False``, keep the first parameter (typically named
   ``self``) of the bound method in a signature.
```

PS: Yeah, and ``bound_arg`` sounds better.

Comment (sobolevn) [reply Doc/library/inspect.rst]:
```python
    @classmethod
    def from_callable(cls, obj, *,
                      follow_wrapped=True, bound_arg=False,
                      globals=None, locals=None, eval_str=False):
        """Constructs Signature for the given callable object."""
        return _signature_from_callable(obj, sigcls=cls,
                                        follow_wrapper_chains=follow_wrapped,
                                        skip_bound_arg=not bound_arg,
                                        globals=globals, locals=locals, eval_str=eval_str)
```

Ok! I wanted to have the same name, but since all other objects do not expose `skip_bound_arg`, it is just an implementation detail. So, we can show the better name.

**Comment by sobolevn** [comment]
Done, now it is named `bound_arg`! 👍

**Comment by encukou** [comment]
What's the motivation for this change? I can't find the discussion behind it.

My worry is that it only works on *some* callables, and it's not very clear which ones those are -- or how to make a custom callable (e.g. one implemented in the C API) work like a method.

**Comment by sobolevn** [comment]
> What's the motivation for this change?

I want to deprecate `getfullargspec`, which works almost like `signature()`, but with this difference. See https://github.com/python/cpython/blob/2c451489122d539080c8d674b391dedc1dedcb53/Lib/inspect.py#L1403-L1423

> My worry is that it only works on some callables, and it's not very clear which ones those are -- or how to make a custom callable (e.g. one implemented in the C API) work like a method.

It will work exactly like this older function.

**Comment by encukou** [comment]
Is this behaviour *useful*, or is it a bug that's kept in `getfullargspec` for backwards compatibility?

File: Doc/library/inspect.rst [hunk 9]
```diff
@@ -675,6 +675,12 @@ function.
    Accepts a wide range of Python callables, from plain functions and classes to
    :func:`functools.partial` objects.
 
+   If *follow_wrapped* is ``False`` *callable* will not be unwrapped
+   (``callable.__wrapped__`` will not be used to unwrap decorated callables).
+
+   If *bound_arg* is ``False``, remove ``self`` parameter
+   from the method signature.
```
Comment (ncoghlan) [inline Doc/library/inspect.rst:682]: It's better to describe the non-default behaviour for boolean options (so folks know what they'll get if they request non-default behaviour):

```suggestion
   If *bound_arg* is ``True``, report the (already bound) first parameter on bound
   instance  or class methods (usually ``self`` or ``cls``). This emulates the historical
   behaviour of the deprecated :func:`getfullargspec` function.
```

(Describing both behaviours is also an option, but it seemed excessively verbose in this case)

File: Doc/whatsnew/3.13.rst [hunk 10]
```diff
@@ -521,6 +521,15 @@ io
   built on debug mode <debug-build>`.
   (Contributed by Victor Stinner in :gh:`62948`.)
 
+inspect
+-------
+
+* Add *bound_arg* parameter to :func:`inspect.Signature.from_callable`
+  and :func:`inspect.signature`: keep the ``self`` parameter
+  in the method signature if *bound_arg* is True.
```
Comment (ncoghlan) [inline Doc/whatsnew/3.13.rst:529]: Reword here to match the updated main function docs (including mentioning the getfullargspec compatibility)

File: Doc/whatsnew/3.13.rst [hunk 11]
```diff
@@ -521,6 +521,15 @@ io
...
+  :pypi:`inspect313` package has a backport of this feature.
```
Comment (ncoghlan) [inline Doc/whatsnew/3.13.rst:530]: Maybe make the backport mention a seealso note in the module docs rather than only putting it here?

File: Misc/NEWS.d/next/Library/2024-03-10-14-45-00.gh-issue-108901.vMlWej.rst [hunk 12]
```diff
@@ -0,0 +1,3 @@
+Add *bound_arg* keyword-only parameter to
+:func:`inspect.Signature.from_callable` and :func:`inspect.signature`.
+If *bound_arg* is ``True``, keep ``self`` parameter in method a signature.
```
Comment (ncoghlan) [inline Misc/NEWS.d/next/Library/2024-03-10-14-45-00.gh-issue-108901.vMlWej.rst:3]: Mention the getfullargspec compatibility aspect here.

**Comment by vstinner** [comment]
> Is this behaviour useful, or is it a bug that's kept in getfullargspec for backwards compatibility?

IMO it's useful in general, not only for backward compatibility. Keep self or not should be configurable.

**Comment by ncoghlan** [comment]
> Is this behaviour _useful_, or is it a bug that's kept in `getfullargspec` for backwards compatibility?

I think it was primarily a quirk of the way `getfullargspec` was originally implemented (i.e. it followed bound methods down to the underlying function to get the signature details, and then just returned those details as is without accounting for the first parameter already being bound).

However, the suggested flag isn't hard to maintain, since the complexity is in the branch that corrects the underlying function signature to remove the already bound first parameter (the default behaviour). The "unwrap bound methods" branch is literally just `return sig`: https://github.com/python/cpython/blob/3989894a396c8c2abc78fcd8758542eed0f68912/Lib/inspect.py#L2517

I do think making the flag public is worth including as a step prior to full programmatic deprecation of `getfullargspec`. With the suggested docs change, I also think the `bound_arg` parameter name is a reasonable one.

**Comment by encukou** [comment]
> the suggested flag isn't hard to maintain

My worry is that while the flag isn't hard to maintain as an *internal detail* of how `getfullargspec` extracts information from Python functions/methods, `signature` should work for other callables, like instances of C extension types (e.g. ones with `Py_TPFLAGS_METHOD_DESCRIPTOR`), or Cython functions. We might not be able to support these things transparently, but there should be a clear way to make custom objects that work with the flag. Or a documented way, at least.

(Unfortunately I'll probably not have time to investigate this worry before beta 1. If the feature is waiting for *my* review, it won't make it.)
